### PR TITLE
[IMP] website_forum: question page

### PR DESCRIPTION
**Pull request by dwa-odoo**
purpose

Misc improvements on Forum

specifications

- Change title of new question page from 'new_question' to 'New Question'.
- Display tooltip  "only one answer per question is allowed" only for question-
  answer type of forum.
- Prevent the user with the warning popup to submit his question or answer
  without description .
- Once the question is flagged, add two options for it:
  'Validate' and 'Mark as Offensive'. So, directly can validate or submit with
  flagged reason from question page.
- Add a new tour for flag a question and show how to validate or submit flagged
  reason of question from question page.

task-2167561
Closes #45352

### CODE REVIEW

**Comment by sbuhl** [comment]
@dwa-odoo May I ask you to separate your PR in several commits.
One per point of the specs. 
Note that the point " 2) While posting question without description, it's showing a warning on a new page, once user post the question. " has been another task, so you can remove it.
So 3 commits. 
Thanks in advance.

File: addons/website_forum/views/website_forum.xml [hunk 1]
```diff
@@ -826,7 +826,7 @@
 </template>
 
 <!-- Edition: ask your question -->
-<template id="new_question">
+<template id="new_question" name="New Question">
```
Comment (sbuhl) [inline addons/website_forum/views/website_forum.xml:829]: Replace into `New Post`

Comment (dwa-odoo) [reply addons/website_forum/views/website_forum.xml]: @sbuhl changes done.

File: addons/website_forum/views/website_forum.xml [hunk 2]
```diff
@@ -953,7 +967,7 @@
             users having a high karma can see offensive posts to moderate
             them.
         </p>
-        <form t-attf-action="/forum/#{ slug(forum) }/#{offensive and 'post' or 'question'}/#{slug(question)}/#{offensive and 'mark_as_offensive' or 'close'}" method="post" role="form" class="mt32 mb64 js_website_submit_form">
+        <form t-attf-action="/forum/#{ slug(forum) }/#{offensive and 'post' or 'question'}/#{slug(question)}/#{offensive and 'mark_as_offensive' or 'close'}" method="post" role="form" t-attf-class="mt32 js_website_submit_form #{'mb64 ' if not dialog else ''}">
```
Comment (rdeodoo) [inline addons/website_forum/views/website_forum.xml:970]: Let's keep it simple, the less code the better.
You can just remove the `mb64` from the code, and so you don't need a `dialog` variable :)
It is not especially ugly without the `mb64` anyway..
image

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Could you also please rebase the branch?
Seems like a very old branch, you can't type in description field in the frontend, so have to create posts through the backend for testing.
After rebase & review processed, I can have another look

**Comment by JKE-be** [comment]
@robodoo delegate=rdeodoo

File: addons/website_forum/static/src/js/tours/website_forum.js [hunk 3]
```diff
@@ -76,4 +76,70 @@ odoo.define("website_forum.tour_forum", function (require) {
...
+        },
+    }, {
+        trigger: "button:contains(\"Post Your Question\")",
+        extra_trigger: "input[id=s2id_autogen2]:not(:propValue(\"Tags\"))",
+        content: _t("Click to post your question."),
+        position: "bottom",
+    }, {
+        extra_trigger: 'div.modal.modal_shown',
+        trigger: ".modal-header button.close",
+        auto: true,
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website_forum/static/src/js/tours/website_forum.js:101]: Still something to be validated by a product owner as not part of the task:
@sbuhl @pierre-pde do we want a tour to show the flagging feature?
If yes, @dwa-odoo either remove the duplicated part of the tour and make it start directly on the post page or either just add those steps at the end of the other tour

File: addons/website_forum/static/src/js/tours/website_forum.js [hunk 4]
```diff
@@ -63,5 +63,72 @@ odoo.define("website_forum.tour_forum", function (require) {
         trigger: ".o_wforum_validate_toggler[data-karma=\"20\"]:first",
         content: _t("Click here to accept this answer."),
         position: "right",
+    }
+    ]);
+
+    tour.register("flag_question", {
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/tours/website_forum.js:69]: This should be called in python to be part of the testing suite..
This will:
1. Strengthen our testing suite
2. Make sure this tour is not broken one day without noticing it

File: addons/website_forum/static/src/js/tours/website_forum.js [hunk 5]
```diff
@@ -63,5 +63,72 @@ odoo.define("website_forum.tour_forum", function (require) {
...
+        content: _t("Click to flag question."),
+        position: "right",
+        run: 'click',
+    }, {
+        trigger: ".input-group-append .js_flag_validator .text-success",
+        content: _t("Click here to validate question."),
+        position: "bottom",
+        run: 'click',
+    }, {
+        trigger: "button:contains(\'Flag\')",
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/tours/website_forum.js:118]:
```suggestion
        trigger: "button:contains('Flag')",
```
What are those backslashes for?

File: addons/website_forum/static/src/js/tours/website_forum.js [hunk 6]
```diff
@@ -63,5 +63,72 @@ odoo.define("website_forum.tour_forum", function (require) {
...
+    }, {
+        extra_trigger: 'div.modal.modal_shown',
+        trigger: ".modal-header button.close",
+        auto: true,
+    }, {
+        trigger: ".o_wforum_post section .dropdown #dropdownMenuLink",
+        content: _t("Click here to show dropdown options."),
+        position: "right",
+    }, {
+        trigger: "button:contains(\"Flag\")",
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/tours/website_forum.js:107]:
```suggestion
        trigger: "button:contains('Flag')",
```
?

File: addons/website_forum/static/src/js/tours/website_forum.js [hunk 7]
```diff
@@ -63,5 +63,72 @@ odoo.define("website_forum.tour_forum", function (require) {
...
+        run: 'click',
+    }, {
+        trigger: "button:contains(\'Flag\')",
+        extra_trigger: ".o_wforum_post section .dropdown.show .dropdown-menu-right.show",
+        content: _t("Click to flag question."),
+        position: "right",
+        run: 'click',
+    }, {
+        trigger: ".input-group-append .js_flag_mark_as_offensive .text-danger",
+        content: _t("Click here to submit mark as offensive question ."),
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/tours/website_forum.js:125]:
```suggestion
        content: _t("Click here to submit mark as offensive question."),
```
This is a tour that user will see, please pay attention to this kind of detail

File: addons/website_forum/static/src/js/tours/website_forum.js [hunk 8]
```diff
@@ -63,5 +63,72 @@ odoo.define("website_forum.tour_forum", function (require) {
...
+        position: "bottom",
+        run: 'click',
+    }, {
+        trigger: ".form-group .btn-danger",
+        content: _t("Click here to submit reason of flagged question."),
+        position: "bottom",
+        run: 'click',
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/tours/website_forum.js:132]:
```suggestion
```
click is the default run no so it is useless no? same every steps

File: addons/website_forum/static/src/js/tours/website_forum.js [hunk 9]
```diff
@@ -63,5 +63,72 @@ odoo.define("website_forum.tour_forum", function (require) {
...
+        content: _t("Give your question title."),
+    }, {
+        trigger: ".note-editable p",
+        extra_trigger: "input[name=post_name]:not(:propValue(\"\"))",
+        content: _t("Put your question here."),
+        position: "bottom",
+        run: "text",
+    }, {
+        trigger: ".select2-choices",
+        extra_trigger: ".note-editable p:not(:containsExact(\"<br>\"))",
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/tours/website_forum.js:87]:
```suggestion
        extra_trigger: ".note-editable p:not(:containsExact('<br>'))",
```

File: addons/website_forum/static/src/js/tours/website_forum.js [hunk 10]
```diff
@@ -63,5 +63,72 @@ odoo.define("website_forum.tour_forum", function (require) {
...
+        content: _t("Insert tags related to your question."),
+        position: "top",
+        run: function (actions) {
+            actions.auto("input[id=s2id_autogen2]");
+        },
+    }, {
+        trigger: "button:contains(\"Post Your Question\")",
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/tours/website_forum.js:94]: What's up with the backslashes?

File: addons/website_forum/views/website_forum.xml [hunk 11]
```diff
@@ -603,7 +617,7 @@
             </span>
             <a t-else="" class="nav-link rounded-pill text-reset" t-attf-href="/forum/#{slug(forum)}/flagged_queue">
                 <i class="fa fa-flag fa-fw"/> Flagged
-                <span t-attf-class="badge pull-right #{forum.count_flagged_posts > 0 and 'badge-danger' or 'badge-light'}" t-esc="forum.count_flagged_posts"/>
+                <span id="count_flagged_posts" t-attf-class="badge pull-right #{forum.count_flagged_posts > 0 and 'badge-danger' or 'badge-light'}" t-esc="forum.count_flagged_posts"/>
```
Comment (rdeodoo) [inline addons/website_forum/views/website_forum.xml:620]: :+1: this was actually broken by https://github.com/odoo/odoo/commit/c8c8eb3d5652e9834d0ecd5aad907ee359edb657 which removed that id by mistake -> the js code couldnt find the element anymore and left some dead JS code

File: addons/website_forum/static/src/js/website_forum.js [hunk 12]
```diff
@@ -494,6 +503,62 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
...
+        const data = $currentTarget.data();
+        this._rpc({
+            model: 'forum.post',
+            method: data.action,
+            args: [data.postId],
+        }).then(result => {
+            if (result) {
+                let $flaggedButton = $currentTarget.parent().toggleClass('d-none').siblings('button'),
+                    $child = $flaggedButton.children(),
+                    $count_flagged_posts = $('span#count_flagged_posts'),
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/website_forum.js:522]:
```suggestion
                    $countFlaggedPosts = $('span#count_flagged_posts'),
```
As per guidelines. Linter should tell you that.
image

File: addons/website_forum/static/src/js/website_forum.js [hunk 13]
```diff
@@ -317,13 +320,19 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
             } else if (data.success) {
                 var elem = $link;
                 if (data.success === 'post_flagged_moderator') {
-                    elem.data('href') && elem.html(' Flagged');
-                    var c = parseInt($('#count_flagged_posts').html(), 10);
-                    c++;
-                    $('#count_flagged_posts').html(c);
+                    let $child = elem.children(),
+                        $count_flagged_posts = $('span#count_flagged_posts'),
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/website_forum.js:324]:
```suggestion
                        $count_flagged_posts = this.$('span#count_flagged_posts'),
```
or `self` I let you check

File: addons/website_forum/static/src/js/website_forum.js [hunk 14]
```diff
@@ -317,13 +320,19 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
...
+                    $count_flagged_posts.removeClass("badge-light");
+                    $count_flagged_posts.addClass("badge-danger");
+                    $count_flagged_posts.html(c);
+                    elem.next('#flag_validator').removeClass('d-none');
                 } else if (data.success === 'post_flagged_non_moderator') {
-                    elem.data('href') && elem.html(' Flagged');
-                    var forumAnswer = elem.closest('.forum_answer');
+                    let $child = elem.children(),
+                        forumAnswer = elem.closest('.forum_answer');
+                    elem.html(' Flagged').prepend($child);
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/website_forum.js:335]: :cry: this is not translated.. I know it was there before but :/
maybe just add a TODO for now, I didn't yet review in depth the business code, so we will tackle this one in next review

File: addons/website_forum/static/src/js/website_forum.js (hunk 13 above)
Comment (rdeodoo) [reply addons/website_forum/static/src/js/website_forum.js:324]: And the `span` part you added is probably useless no? it's an ID anyway, there shouldn't be more than one element..

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Check runbot after you push when you can :) in this case you can see there is error coming from this PR
Also, after you process the review, please either:
- move the task to RD Dev Green
- re-request a review on the PR from me
- ping me on the PR
So I/we know this should be reviewed again, otherwise we can't really know :)

File: addons/website_forum/views/website_forum.xml (hunk 11 above)
Comment (rdeodoo) [reply addons/website_forum/views/website_forum.xml:620]: You can remove this from your PR, I backported it in the version it was broken...
https://github.com/odoo/odoo/pull/74796

File: addons/website_forum/views/website_forum.xml [hunk 15]
```diff
@@ -1311,7 +1332,7 @@
                         <div t-if="question.uid_has_answered and answer.create_uid.id == uid"
                             class="btn-group btn-group-sm mr-1 mr-md-2">
                             <a class="btn btn-sm px-2 btn-primary"
-                                title="Only one answer per question is allowed"
+                                t-attf-title="#{'Only one answer per question is allowed' if question.forum_id.mode == 'questions' else ''}"
```
Comment (rdeodoo) [inline addons/website_forum/views/website_forum.xml:1335]:
```suggestion
                                t-att-title="'Only one answer per question is allowed' if question.forum_id.mode == 'questions' else ''"
```
:thinking:

File: addons/website_forum/static/src/js/website_forum.js [hunk 16]
```diff
@@ -292,43 +295,45 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
...
-                    elem.data('href') && elem.html(' Flagged');
-                    var c = parseInt($('#count_flagged_posts').html(), 10);
-                    c++;
-                    $('#count_flagged_posts').html(c);
-                } else if (data.success === 'post_flagged_non_moderator') {
-                    elem.data('href') && elem.html(' Flagged');
-                    var forumAnswer = elem.closest('.forum_answer');
-                    forumAnswer.fadeIn(1000);
-                    forumAnswer.slideUp(1000);
-                }
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/website_forum.js:329]: All this diff for literally only 3 lines of real diff..
(you can see it in `?w=1` in the URL)
I don't like such big diff that just do nothing but why not.. I would have at least put it in a separate commit "Clean code before modify it" but whatever

File: addons/website_forum/static/src/js/website_forum.js [hunk 17]
```diff
@@ -292,43 +295,45 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
...
+                message: message,
+                title: _t("Access Denied"),
+                sticky: false,
+                type: "warning",
+            });
+        } else if (data.success) {
+            const child = elem.firstElementChild;
+            if (data.success === 'post_flagged_moderator') {
+                const countFlaggedPosts = this.el.querySelector('#count_flagged_posts');
+                elem.innerText = _t(' Flagged');
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/website_forum.js:323]: The previous condition "if href" isn't needed anymore?

File: addons/website_forum/static/src/js/website_forum.js [hunk 18]
```diff
@@ -292,43 +295,45 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
...
-                } else if (data.success === 'post_flagged_non_moderator') {
-                    elem.data('href') && elem.html(' Flagged');
-                    var forumAnswer = elem.closest('.forum_answer');
-                    forumAnswer.fadeIn(1000);
-                    forumAnswer.slideUp(1000);
-                }
+        const elem = ev.currentTarget;
+        const data = await this._rpc({
+            route: elem.dataset.href || (elem.getAttribute('href') !== '#' && elem.getAttribute('href')) || elem.closest('form').getAttribute('action'),
+        })
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/website_forum.js:303]: missing `;`, linter should tell you that

File: addons/website_forum/static/src/js/website_forum.js [hunk 19]
```diff
@@ -494,6 +499,63 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
         $('.forum_intro').slideUp();
         return true;
     },
+    /**
+     * @private
+     * @param {Event} ev
+     */
+    _onFlagValidatorClick: async function (ev) {
+        ev.preventDefault();
+        const currentTarget = ev.currentTarget
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/website_forum.js:508]: missing `;` please check your linter

File: addons/website_forum/static/src/js/website_forum.js [hunk 20]
```diff
@@ -494,6 +499,63 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
...
+            args: [parseInt(currentTarget.dataset.postId)],
+        });
+        if (result) {
+            currentTarget.parentElement.classList.toggle('d-none');
+            const flaggedButton = currentTarget.parentElement.previousElementSibling,
+                child = flaggedButton.firstElementChild,
+                countFlaggedPosts = this.el.querySelector('#count_flagged_posts'),
+                count = parseInt(countFlaggedPosts.innerText, 10) - 1;
+
+            flaggedButton.innerText = _t(' Flag')
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/website_forum.js:521]: `;` ..

File: addons/website_forum/static/src/js/website_forum.js [hunk 21]
```diff
@@ -494,6 +499,63 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
...
+    /**
+     * @private
+     * @param {Event} ev
+     */
+    _onFlagMarkAsOffensiveClick: async function (ev) {
+        ev.preventDefault();
+        const result = await this._rpc({
+            route: $(ev.currentTarget).data('action'),
+            params: {}
+        })
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/website_forum.js:538]: again..

File: addons/website_forum/views/website_forum.xml [hunk 22]
```diff
@@ -4,13 +4,27 @@
 
 <!-- Editor custom -->
 <!-- helper -->
+<template id="show_flag_validator">
+    <div id="flag_validator" t-attf-class="input-group-append #{'ml-1' if object.state == 'flagged' else 'd-none'}">
+        <a t-att-data-post-id="object.id" t-attf-data-action="validate" class="badge badge-light mt-2 mb-2 mr-2 o_wforum_flag_validator" title="Validate" data-toggle="tooltip">
+            <i class="fa fa-check text-success"/>
+        </a>
+        <a t-attf-data-action="/forum/#{object.id}/ask_for_mark_as_offensive" href="#" class="badge badge-light mt-2 mb-2 mr-2 o_wforum_flag_mark_as_offensive" title="Mark as Offensive"  data-toggle="tooltip">
+            <i class="fa fa-times text-danger "/>
```
Comment (rdeodoo) [inline addons/website_forum/views/website_forum.xml:13]: extra space

File: addons/website_forum/controllers/main.py [hunk 23]
```diff
@@ -526,17 +538,16 @@ def post_flag(self, forum, post, **kwargs):
...
-        values = self._prepare_user_values(forum=forum)
-        values.update({
-            'question': post,
-            'forum': forum,
-            'reasons': offensive_reasons,
-            'offensive': True,
-        })
+    @http.route('/forum/<model("forum.forum"):forum>/post/<model("forum.post"):post>/ask_for_mark_as_offensive', type='http', auth="user", methods=['GET'], website=True)
+    def post_http_ask_for_mark_as_offensive(self, forum, post, **kwargs):
+        values = self.ask_for_mark_as_offensive(post, **kwargs)
```
Comment (rdeodoo) [inline addons/website_forum/controllers/main.py:550]: isn't this missing the post.can_moderate check?

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Functionally, it does what the task ask.

File: addons/website_forum/controllers/main.py (hunk 23 above)
Comment (jpr-odoo) [reply addons/website_forum/controllers/main.py:553]: Done!

File: addons/website_forum/static/src/js/website_forum.js (hunk 17 above)
Comment (jpr-odoo) [reply addons/website_forum/static/src/js/website_forum.js]: we couldn't find any such use case, that's why we have removed it

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Can you fix this one then it should be good to me

File: addons/website_forum/views/website_forum.xml [hunk 24]
```diff
@@ -1254,11 +1273,13 @@
             </t>
             <t t-if="question.active" href="#" t-call="website_forum.link_button">
                 <t t-set="url" t-value="'/forum/' + slug(forum) +'/post/' + slug(question) + '/flag'"/>
-                <t t-set="label">Flag</t>
+                <t t-set="label" t-value="question.state == 'flagged' and 'Flagged' or 'Flag'"/>
```
[the remark this answers is not in the document]
Comment (jpr-odoo) [reply addons/website_forum/views/website_forum.xml:1276]: Okey

File: addons/website_forum/controllers/main.py [hunk 25]
```diff
@@ -526,17 +539,18 @@ def post_flag(self, forum, post, **kwargs):
...
 
-    @http.route('/forum/<model("forum.forum"):forum>/post/<model("forum.post"):post>/ask_for_mark_as_offensive', type='http', auth="user", methods=['GET'], website=True)
-    def post_ask_for_mark_as_offensive(self, forum, post, **kwargs):
-        offensive_reasons = request.env['forum.post.reason'].search([('reason_type', '=', 'offensive')])
+    @http.route('/forum/<model("forum.post"):post>/ask_for_mark_as_offensive', type='json', auth="user", methods=['POST'], website=True)
+    def post_json_ask_for_mark_as_offensive(self, post, **kwargs):
+        if not post.can_moderate:
+            return {'error': _('%d karma required to mark a post as offensive.') % post.forum_id.karma_moderate}
+        values = self.ask_for_mark_as_offensive(post, **kwargs)
+        return {'mark_as_offensive' : request.env['ir.ui.view']._render_template('website_forum.mark_as_offensive', values)}
```
Comment (rdeodoo) [inline addons/website_forum/controllers/main.py:547]:
```suggestion
        return {'mark_as_offensive': request.env['ir.ui.view']._render_template('website_forum.mark_as_offensive', values)}
```
linter.. (by the way I am not supposed to be your linter)

Comment (jpr-odoo) [reply addons/website_forum/controllers/main.py:547]: really it's bad :-1: . something wrong with linter, I think need to re-install editor's all thinks to at my side

File: addons/website_forum/static/src/js/website_forum.js [hunk 26]
```diff
@@ -501,6 +508,62 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
...
+     * @private
+     * @param {Event} ev
+     */
+    _onFlagMarkAsOffensiveClick: async function (ev) {
+        ev.preventDefault();
+        const result = await this._rpc({
+            route: $(ev.currentTarget).data('action'),
+        });
+        if (result && result.error) {
+            this.call('crash_manager', 'show_warning', {
```
[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website_forum/static/src/js/website_forum.js:548]: And.. I actually tried your code because why not and..
Boom :boom: traceback

image

Please, retest all the flows of this PR before saying "done" after this review.

I get this task is probably a task that was done before owl, and it was working at some point, but we already had a few reviews these last few days and you should have tried such flows after rebasing a x months old PR.
That's the bare minimum when sending a PR to review:
1. Check your code is linted ok
2. Check your code works

Note that call `crash_manager` shoudln't have been written in the first place, we always have called displayNotification

File: addons/website_forum/views/website_forum.xml [hunk 27]
```diff
@@ -1254,11 +1273,14 @@
...
                 <t t-set="url" t-value="'/forum/' + slug(forum) +'/post/' + slug(question) + '/flag'"/>
-                <t t-set="label">Flag</t>
+                <t t-if="question.state == 'flagged'" t-set="label">Flagged</t>
+                <t t-else="" t-set="label">Flag</t>
                 <t t-set="inDropdown" t-value="True"/>
                 <t t-set="icon" t-value="not question.can_flag and 'fa-flag-o' or 'fa-flag'"/>
                 <t t-set="karma" t-value="not question.can_flag and question.forum_id.karma_flag or 0"/>
-                <t t-set="classes" t-value="'flag'"/>
+                <t t-set="classes" t-valuef="o_wforum_flag"/>
+                <t t-set="flag_validator" t-value="user.karma >= question.forum_id.karma_moderate"/>
```
[the remark this answers is not in the document]
Comment (jpr-odoo) [reply addons/website_forum/views/website_forum.xml:1282]: Yes, we can. But the same condition is used everywhere in the file to check moderate access. that's why we have used the same for here. if you want then we can change!

[the remark this answers is not in the document]
Comment (rdeodoo) [reply addons/website_forum/views/website_forum.xml:1282]: > But the same condition is used everywhere in the file to check moderate access

Doesn't mean you have to keep writing the same code. And it's not `everywhere`, there is 5 occurrences of `user.karma >= forum.karma_moderate` (incorrect) and 2 occurrences of `question.can_moderate` (correct).

Then if you blame the lines, you will see the incorrect ones have been introduced by a designer during a revamp, which is "normal" as they obviously dont know the business code and helper as we, developers, do.

File: addons/website_forum/static/src/js/website_forum.js [hunk 28]
```diff
@@ -501,6 +510,57 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
...
+     */
+    _onFlagValidatorClick: async function (ev) {
+        ev.preventDefault();
+        const currentTarget = ev.currentTarget;
+        const result = await this._rpc({
+            model: 'forum.post',
+            method: currentTarget.dataset.action,
+            args: [parseInt(currentTarget.dataset.postId)],
+        });
+        if (result) {
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/website_forum.js:525]: How could result be falsy? Either it crashes and it handled at lower level (which shows a toaster) or it will just return something

File: addons/website_forum/static/src/js/website_forum.js [hunk 29]
```diff
@@ -501,6 +510,57 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
...
+    /**
+     * @private
+     * @param {Event} ev
+     */
+    _onFlagMarkAsOffensiveClick: async function (ev) {
+        ev.preventDefault();
+        const result = await this._rpc({
+            route: $(ev.currentTarget).data('action'),
+        });
+        if (result) {
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/website_forum.js:549]: How could result be falsy? Either it crashes and it handled at lower level (which shows a toaster) or it will just return something

+ as I asked, name if `template` not `result`

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Just some minor stuff, can you quickly change it? I will move your task to VALIDATION BE Green as it won't change anything, don't put it back in RD DEV, just push the 2 line diff :)

File: addons/website_forum/static/src/js/website_forum.js [hunk 30]
```diff
@@ -501,6 +510,57 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
...
+                countFlaggedPosts.classList.add("badge-light");
+            }
+            countFlaggedPosts.innerText = count;
+        }
+    },
+    /**
+     * @private
+     * @param {Event} ev
+     */
+    _onFlagMarkAsOffensiveClick: async function (ev) {
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/website_forum.js:544]:
```suggestion
    async _onFlagMarkAsOffensiveClick(ev) {
```

File: addons/website_forum/static/src/js/website_forum.js [hunk 31]
```diff
@@ -501,6 +510,57 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
         $('.forum_intro').slideUp();
         return true;
     },
+    /**
+     * @private
+     * @param {Event} ev
+     */
+    _onFlagValidatorClick: async function (ev) {
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/website_forum.js:517]:
```suggestion
    async _onFlagValidatorClick(ev) {
```

File: addons/website_forum/static/src/js/website_forum.js [hunk 32]
```diff
@@ -501,6 +510,53 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
...
         return true;
     },
+    /**
+     * @private
+     * @param {Event} ev
+     */
+    async _onFlagValidatorClick(ev) {
+        ev.preventDefault();
+        const currentTarget = ev.currentTarget;
+        const result = await this._rpc({
```
Comment (rdeodoo) [inline addons/website_forum/static/src/js/website_forum.js:520]: unused variable, linter pls :)

**Review by rdeodoo — COMMENTED** [review]
Can you please rebase it? We will try to merge it asap

**Comment by rdeodoo** [comment]
Also, when clicking on Flag in the dropdown and the cross/check appear, the dropdown item remain gray (active class behavior), can you check to remove that? it is quite ugly as JKE pointed out to me yesterday

File: addons/website_forum/controllers/main.py [hunk 33]
```diff
@@ -526,17 +539,18 @@ def post_flag(self, forum, post, **kwargs):
             return {'error': 'anonymous_user'}
         return post.flag()[0]
 
-    @http.route('/forum/<model("forum.forum"):forum>/post/<model("forum.post"):post>/ask_for_mark_as_offensive', type='http', auth="user", methods=['GET'], website=True)
-    def post_ask_for_mark_as_offensive(self, forum, post, **kwargs):
-        offensive_reasons = request.env['forum.post.reason'].search([('reason_type', '=', 'offensive')])
+    @http.route('/forum/<model("forum.post"):post>/ask_for_mark_as_offensive', type='json', auth="user", methods=['POST'], website=True)
```
Comment (JKE-be) [inline addons/website_forum/controllers/main.py:542]: methods is useless with JSON.

File: addons/website_forum/static/src/js/website_forum.js [hunk 34]
```diff
@@ -302,9 +305,9 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
     _onFlagAlertClick: function (ev) {
         var self = this;
         ev.preventDefault();
-        var $link = $(ev.currentTarget);
+        const elem = ev.currentTarget;
         this._rpc({
-            route: $link.data('href') || ($link.attr('href') !== '#' && $link.attr('href')) || $link.closest('form').attr('action'),
+            route: elem.dataset.href || (elem.getAttribute('href') !== '#' && elem.getAttribute('href')) || elem.closest('form').getAttribute('action'),
```
Comment (JKE-be) [inline addons/website_forum/static/src/js/website_forum.js:310]: free diff ?

File: addons/website_forum/static/src/js/website_forum.js [hunk 35]
```diff
@@ -315,22 +318,28 @@ publicWidget.registry.websiteForum = publicWidget.Widget.extend({
                 } else if (data.error === 'post_non_flaggable') {
                     message = _t("This post can not be flagged");
                 }
-                this.displayNotification({
+                self.displayNotification({
```
Comment (JKE-be) [inline addons/website_forum/static/src/js/website_forum.js:321]: use this and es6? `data =>`

File: addons/website_forum/views/website_forum.xml [hunk 36]
```diff
@@ -1311,7 +1333,7 @@
                         <div t-if="question.uid_has_answered and answer.create_uid.id == uid"
                             class="btn-group btn-group-sm mr-1 mr-md-2">
                             <a class="btn btn-sm px-2 btn-primary"
-                                title="Only one answer per question is allowed"
+                                t-att-title="'Only one answer per question is allowed' if question.forum_id.mode == 'questions' else ''"
```
Comment (JKE-be) [inline addons/website_forum/views/website_forum.xml:1336]: not translatable now

File: addons/website_forum/views/website_forum.xml [hunk 37]
```diff
@@ -4,13 +4,27 @@
 
 <!-- Editor custom -->
 <!-- helper -->
+<template id="show_flag_validator">
+    <div id="flag_validator" t-attf-class="input-group-append #{'ml-1' if object.state == 'flagged' else 'd-none'}">
+        <a t-att-data-post-id="object.id" t-attf-data-action="validate" class="badge badge-light mt-2 mb-2 mr-2 o_wforum_flag_validator" title="Validate" data-toggle="tooltip">
```
Comment (JKE-be) [inline addons/website_forum/views/website_forum.xml:9]: mt-2 + mb-2 == my-2

**Review by JKE-be — COMMENTED** [review]
nice... some small changes fafter a first look. Not run the branch

File: addons/website_forum/static/src/js/website_forum.js (hunk 35 above)
Comment (rdeodoo) [reply addons/website_forum/static/src/js/website_forum.js:321]: @JKE-be Actually this line should be fixed with your PR (this is the bug we discussed yesterday introduced by https://github.com/odoo/odoo/commit/57ba855b66f2ff1d393f639bbc18dd444e2afa8f)
So this diff should be removed from this PR once bug is fixed

**Comment by rdeodoo** [comment]
By the way you can remove the PR references from your commit messages, it will prevent all the push force messages, and is correctly added now on every commit by the mergebot on multi-commit PR :)

**Comment by JKE-be** [comment]
@robodoo delegate=rdeodoo

**Comment by rdeodoo** [comment]
@robodoo r+
